source/opt/decoration_manager.cpp

```diff
@@ -231,29 +231,78 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, true)) {
+    switch (inst->opcode()) {
+      case SpvOpDecorate:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
+    switch (inst->opcode()) {
+      case SpvOpNop: break;
+      case SpvOpGroupDecorate: {
+        // replace decoration instruction with one containing the new id as well.
+        std::vector<ir::Operand> new_operands;
+        new_operands.insert(inst->begin(), inst->begin(), inst->end());
+        new_operands.push_back(ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID, {to}));
+        std::unique_ptr<ir::Instruction> new_inst(new ir::Instruction(inst->opcode(), inst->type_id(), inst->result_id(), new_operands));
+        id_to_decoration_insts_[from].push_back(new_inst.get());
```

Comment:
I expected `to` as the map index.  ??

Reply:
Also, for all IDs x in the in-operand-list of inst, id_to_decoration_insts_[x] incorrectly contains &inst.  That stale reference must be removed.

Reply:
Yes...
Had it right, before, with AddOperand(). If this function is okay for you, I'll do it by adding the operands and def_use_mgr as well as dec_mgr remain up to date.